### PR TITLE
(BOLT-636) Add logger for manifest block compilation

### PR DESCRIPTION
**Pull request by MikaelSmith**
Adds a logger to handle messages from compiling manifest blocks for
targets that gathers all log messages created during the compilation and
redirects them to Bolt's logging facilities.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
Comparison of failing compilation (with an apply block containing `fail('now')`).

Before:
```
Starting: apply catalog on localhost:20022
Finished: apply catalog with 1 failure in 1.46 sec
[
  {
    "node": "localhost:20022",
    "status": "failure",
    "result": {
      "_error": {
        "kind": "bolt/apply-error",
        "msg": "Apply failed to compile for Target('localhost:20022', {\"connect-timeout\"=>10, \"tty\"=>false, \"host-key-check\"=>false, \"run-as\"=>\"root\", \"sudo-password\"=>\"bolt\"}): /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/functions/fail.rb:3:in `block in <top (required)>': Evaluation Error: Error while evaluating a Function Call, now (file: /Users/michaelsmith/puppetlabs/bolt/tmp/foo/plans/simple.pp, line: 7, column: 5) (Puppet::PreformattedError)\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/functions.rb:206:in `block (2 levels) in newfunction'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler.rb:51:in `profile'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/functions.rb:199:in `block in newfunction'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/tmp/foo/plans/simple.pp:7:in `stack'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/puppet_stack.rb:33:in `eval'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/puppet_stack.rb:33:in `stack'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/runtime3_support.rb:314:in `call_function'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:976:in `call_function_with_block'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:945:in `eval_CallNamedFunctionExpression'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:48:in `block in visit_this'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:42:in `each'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:42:in `visit_this'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:92:in `visit_this_1'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:81:in `evaluate'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:1282:in `block in unfold'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:1277:in `each'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:1277:in `unfold'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:646:in `eval_LiteralList'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:48:in `block in visit_this'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:42:in `each'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:42:in `visit_this'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:92:in `visit_this_1'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:81:in `evaluate'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/parser/evaluating_parser.rb:63:in `evaluate'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/ast/pops_bridge.rb:125:in `evaluate'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/ast.rb:30:in `safeevaluate'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/resource/type.rb:136:in `evaluate_code'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/resource.rb:78:in `block in evaluate'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler.rb:51:in `profile'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/resource.rb:70:in `evaluate'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:50:in `evaluate_main'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/compiler.rb:165:in `block (2 levels) in compile'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler.rb:51:in `profile'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/compiler.rb:165:in `block in compile'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/context.rb:65:in `override'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet.rb:251:in `override'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/compiler.rb:154:in `compile'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:83:in `compile_node'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:118:in `block (3 levels) in compile_catalog'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/context.rb:65:in `override'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet.rb:251:in `override'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:117:in `block (2 levels) in compile_catalog'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet_pal.rb:699:in `block in in_environment_context'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/context.rb:65:in `override'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet.rb:251:in `override'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet_pal.rb:691:in `in_environment_context'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet_pal.rb:593:in `in_tmp_environment'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:108:in `block in compile_catalog'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:68:in `block in with_puppet_settings'\n\tfrom /Users/michaelsmith/.rbenv/versions/2.3.4/lib/ruby/2.3.0/tmpdir.rb:89:in `mktmpdir'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:59:in `with_puppet_settings'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:105:in `compile_catalog'\n\tfrom /Users/michaelsmith/puppetlabs/bolt/libexec/bolt_catalog:32:in `<main>'\n",
        "details": {
        }
      }
    }
  }
]
```
After:
```
Starting: apply catalog on localhost:20022
/Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/functions/fail.rb:3:in `block in <top (required)>': Evaluation Error: Error while evaluating a Function Call, now (file: /Users/michaelsmith/puppetlabs/bolt/tmp/foo/plans/simple.pp, line: 7, column: 5) (Puppet::PreformattedError)
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/functions.rb:206:in `block (2 levels) in newfunction'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler.rb:51:in `profile'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/functions.rb:199:in `block in newfunction'
	from /Users/michaelsmith/puppetlabs/bolt/tmp/foo/plans/simple.pp:7:in `stack'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/puppet_stack.rb:33:in `eval'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/puppet_stack.rb:33:in `stack'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/runtime3_support.rb:314:in `call_function'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:976:in `call_function_with_block'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:945:in `eval_CallNamedFunctionExpression'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:48:in `block in visit_this'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:42:in `each'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:42:in `visit_this'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:92:in `visit_this_1'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:81:in `evaluate'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:1282:in `block in unfold'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:1277:in `each'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:1277:in `unfold'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:646:in `eval_LiteralList'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:48:in `block in visit_this'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:42:in `each'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:42:in `visit_this'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/visitor.rb:92:in `visit_this_1'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/evaluator/evaluator_impl.rb:81:in `evaluate'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/pops/parser/evaluating_parser.rb:63:in `evaluate'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/ast/pops_bridge.rb:125:in `evaluate'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/ast.rb:30:in `safeevaluate'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/resource/type.rb:136:in `evaluate_code'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/resource.rb:78:in `block in evaluate'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler.rb:51:in `profile'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/resource.rb:70:in `evaluate'
	from /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog/compiler.rb:44:in `evaluate_main'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/compiler.rb:165:in `block (2 levels) in compile'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/util/profiler.rb:51:in `profile'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/compiler.rb:165:in `block in compile'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/context.rb:65:in `override'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet.rb:251:in `override'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/parser/compiler.rb:154:in `compile'
	from /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:42:in `compile_node'
	from /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:77:in `block (3 levels) in compile_catalog'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/context.rb:65:in `override'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet.rb:251:in `override'
	from /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:76:in `block (2 levels) in compile_catalog'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet_pal.rb:699:in `block in in_environment_context'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet/context.rb:65:in `override'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet.rb:251:in `override'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet_pal.rb:691:in `in_environment_context'
	from /Users/michaelsmith/puppetlabs/bolt/vendored/puppet/lib/puppet_pal.rb:593:in `in_tmp_environment'
	from /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:67:in `block in compile_catalog'
	from /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:27:in `block in with_puppet_settings'
	from /Users/michaelsmith/.rbenv/versions/2.3.4/lib/ruby/2.3.0/tmpdir.rb:89:in `mktmpdir'
	from /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:15:in `with_puppet_settings'
	from /Users/michaelsmith/puppetlabs/bolt/lib/bolt/catalog.rb:64:in `compile_catalog'
	from /Users/michaelsmith/puppetlabs/bolt/libexec/bolt_catalog:32:in `<main>'
Finished: apply catalog with 1 failure in 1.43 sec
[
  {
    "node": "localhost:20022",
    "status": "failure",
    "result": {
      "_error": {
        "kind": "bolt/apply-error",
        "msg": "Apply failed to compile for localhost:20022",
        "details": {
        }
      }
    }
  }
]
```

**Comment by donoghuc** [comment]
This is helpful. Learned that `The function 'hiera' is deprecated in favor of using 'lookup'`! 
The `(file & line not available)` seems to be generated by the puppet logger. I guess that would be meaningless information for an apply block anyway. 
```
Starting: apply catalog on sfb0jyfqgekbq2z.delivery.puppetlabs.net
sfb0jyfqgekbq2z.delivery.puppetlabs.net: The function 'hiera' is deprecated in favor of using 'lookup'. See https://puppet.com/docs/puppet/6.0/deprecated_language.html
   (file & line not available)
Finished: apply catalog with 0 failures in 2.78 sec
```

File: spec/fixtures/apply/basic/plans/error.pp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+plan basic::error(TargetSpec $nodes) {
+  return apply($nodes) {
+    debug('Debugging')
+    info('Meh')
+    notice('Helpful')
+    warning('Warned')
+    err('Fire')
+    alert('Stop')
+    crit('Drop')
+    emerg('Roll')
```
Comment (lucywyman) [inline spec/fixtures/apply/basic/plans/error.pp]: Haha :)

File: lib/bolt/applicator.rb [hunk 2]
```diff
@@ -53,15 +56,36 @@ def compile(target, ast, plan_vars)
...
+      raise(ApplyError, target.uri) unless stat.success?
       JSON.parse(out)
     end
 
     def validate_hiera_config(hiera_config)
       if File.exist?(File.path(hiera_config))
         data = File.open(File.path(hiera_config), "r:UTF-8") { |f| YAML.safe_load(f.read) }
         unless data['version'] == 5
-          raise ApplyError.new("All Targets", "Hiera v5 is required.")
+          raise Bolt::ParseError, "Hiera v5 is required, found v#{data['version'] || 3} in #{hiera_config}"
```
[the remark this answers is not in the document]
Comment (MikaelSmith) [reply lib/bolt/applicator.rb]: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/lookup/hiera_config.rb#L152-L154

**Pull request merged**